Coverart popup fixes

poelzi opened the pull request:
Small fixes to the coverart window popup.

* If the window is closed from the window manager, it's internal state gets destroyed.
Reopening the window only shows a empty window and error codes:

`warning [Main] QXcbConnection: XCB error: 8 (BadMatch), sequence: 32110, resource id: 68218659, major code: 130 (Unknown), minor code: 3
`

* Place the window in the center of the mixxx window, not on the center of the first screen.

poelzi committed:
Don't destroy window on close event

If the window is closed via os side, the
widget is partially destroyed. This prevents future requests
to show a empty gray window.

Prevent the window from being destroyed on close() events.

poelzi committed:
Use the screen of the mixxx window for placing dialog window.

The window is always opened on the first screen, not the screen of
the mixxx window. Use the mixxx window as a reference for placement.

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -46,6 +46,12 @@ DlgCoverArtFullSize::DlgCoverArtFullSize(QWidget* parent, BaseTrackPlayer* pPlay
 
+void DlgCoverArtFullSize::closeEvent(QCloseEvent* event) {
+    // prevent the window from being destroyed
+    hide();
```
`slotLoadTrack(nullptr);`

Otherwise a reference to the current `Track` object is kept and connections remain active.

Resetting the track pointer currently does not reset the pixmap. Not sure how to handle this.

uklotzde reviewed (changes requested):
The internal smart pointer to the `Track` object is not released until Mixxx is closed and metadata. The internal state of the dialog has to be reset instead of only hiding the window.

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -46,6 +46,12 @@ DlgCoverArtFullSize::DlgCoverArtFullSize(QWidget* parent, BaseTrackPlayer* pPlay
 }
 
+void DlgCoverArtFullSize::closeEvent(QCloseEvent* event) {
+    // prevent the window from being destroyed
```
Please also note here why this workaround is required, instead of only in the description of this PR.

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -144,15 +150,20 @@ void DlgCoverArtFullSize::slotCoverFound(
     }
     else {
-        availableScreenGeometry = pScreen->availableGeometry();
+        availableScreenGeometry = pScreen->geometry();
```
Keeping the prefix "available" is inconsistent.

poelzi committed:
Clear references on window close

better variable naming

uklotzde commented:
You are still using the whole screen's geometry for calculating the placement, not the window geometry as mentioned in your description. I guess this is intentional, because we are now using the actual screen and not just the primary screen.

uklotzde commented:
...we have to keep using the whole screen. Centering the window inside the `DlgTrackInfo` wouldn't make sense.

uklotzde commented:
Good catch. I noticed the XCB error logs but was not aware of who was causing them.

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -46,6 +46,14 @@ DlgCoverArtFullSize::DlgCoverArtFullSize(QWidget* parent, BaseTrackPlayer* pPlay
 }
 
+void DlgCoverArtFullSize::closeEvent(QCloseEvent* event) {
+    // Since the same instance if opened again by the same parent widget
```
Do we need to handle the case if DlgCoverArtFullSize does not have a parent? In this case it should be closed.

poelzi replied:
> ...we have to keep using the whole screen. Centering the window inside the `DlgTrackInfo` wouldn't make sense.

It centers now the window on the screen in which the preferences window is currently open. This is the expected behaviour in my opinion.

uklotzde replied:
Yes, that's what I mentioned. It works as intended. Only the description of this PR is slightly misleading.

uklotzde replied:
The wording "in the center of the mixxx window" wrong. "in the center of the screen where the parent window is displayed" would be correct.

ronso0 commented:
@poelzi There is are related issues with the fullsize cover and I hoped it was solved by your changes, but unfortunately it's still there. I'd appreciate if you could tackle those here, too.
**1** (minor) zoom into cover, close it. re-open: for an instant the cover is shown with the previous size, then resizes to default
**2** enlarge (zoom in) cover until it's wider than the screen, close it. re-open: it's shown with previous (over)size and can't be scaled down as long as the track is loaded.

While earlier I tried to make it use the previous size and position (no matter if it's the same track) this behaviour is a bug but can also be used to create 'art' https://mixxx.zulipchat.com/#narrow/stream/109122-general/topic/psych.20cover.20glitch
![image](https://user-images.githubusercontent.com/5934199/97978573-0805d300-1dce-11eb-8546-1d81f310549c.png)

poelzi committed:
Cleanup children if there is no parent set.

poelzi committed:
Fix bug where window is opened too large

1. enlarge (zoom in) cover until it's wider than the screen
2. close it.
3. re-open: it's shown with previous (over)size and can't be scaled down as long as the track is loaded.

poelzi committed:
better wording

poelzi commented:
@ronso0 fixed.

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -46,10 +46,27 @@ DlgCoverArtFullSize::DlgCoverArtFullSize(QWidget* parent, BaseTrackPlayer* pPlay
 }
 
+void DlgCoverArtFullSize::closeEvent(QCloseEvent* event) {
+    hide();
```
Why are the first two lines executed unconditionally?

It works as intended and the redundancy doesn't really matter. I am just asking questions that others might have in the future.

poelzi replied:
We are not using the case without a parent in the current code base. It is just there in case we will use it, it does the right thing. Currently there is a VERIFY_OR_DEBUG_ASSERT for the parentless case in init anyway.

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -46,10 +46,27 @@ DlgCoverArtFullSize::DlgCoverArtFullSize(QWidget* parent, BaseTrackPlayer* pPlay
     if (!pTrack) {
         return;
     }
+    // The real size will be calculated later.
```
This is exactly what comments are good for :+1:

uklotzde commented on src/library/dlgcoverartfullsize.cpp:
```diff
@@ -144,22 +163,27 @@ void DlgCoverArtFullSize::slotCoverFound(
+    QWidget* parent = parentWidget();
+    VERIFY_OR_DEBUG_ASSERT(parent) {
+        qWarning() << "DlgCoverArtFullSize does not have a parent.";
+        parent = this;
```
The name of the local variable is misleading when reusing it for this. If you don't know a better name then just use `parentOrThis`.

poelzi committed:
cleanup naming. Don't run redunant code on close

ronso0 commented:
> @ronso0 fixed. For some reason setGeometry does not resize the window despite the fact the provided QRect is correct.
> Setting a small size before show is called, fixes the problem.

Perfect, thanks!

poelzi commented:
@uklotzde ready for merge

uklotzde reviewed (approved):
Thank you! LGTM
